aws-storage-s3/src/androidTest/java/com/amplifyframework/storage/s3/AWSS3StorageUploadTest.java

```diff
@@ -0,0 +1,238 @@
+                Float progress = progressEvent.getData();
+                if (progress != null && progress > 0) {
+                    op.cancel();
+                }
+            }
+        });
+        subscriptions.add(progressToken);
+
+        // Listen to Hub events for cancel
+        SubscriptionToken cancelToken = Amplify.Hub.subscribe(HubChannel.STORAGE, hubEvent -> {
+            if ("uploadState".equals(hubEvent.getName())) {
+                HubEvent<String> stateEvent = (HubEvent<String>) hubEvent;
+                TransferState state = TransferState.getState(stateEvent.getData());
+                if (TransferState.CANCELED.equals(state)) {
+                    canceled.countDown();
+                }
+            }
+        });
+        subscriptions.add(cancelToken);
+
+        // Assert that the required conditions have been met
+        assertTrue(canceled.await(EXTENDED_TIMEOUT_IN_SECONDS, TimeUnit.SECONDS));
+    }
+
+    /**
+     * Tests that file upload operation can be paused and resumed
+     * while the transfer hasn't completed yet.
+     *
+     * @throws Exception if upload is not paused, resumed, and
+     *         completed successfully before timeout
+     */
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testUploadFileIsResumable() throws Exception {
+        final CountDownLatch completed = new CountDownLatch(1);
+        final CountDownLatch resumed = new CountDownLatch(1);
+
+        // Begin uploading large file
+        StorageUploadFileOperation<?> op = Amplify.Storage.uploadFile(
+            largeFile.getName(),
+            largeFile.getAbsolutePath(),
+            options,
+            onResult -> completed.countDown(),
+            onError -> fail("Upload is not successful.")
+        );
+
+        // Listen to Hub events to pause when progress has been made
+        SubscriptionToken pauseToken = Amplify.Hub.subscribe(HubChannel.STORAGE, hubEvent -> {
+            if ("uploadProgress".equals(hubEvent.getName())) {
+                HubEvent<Float> progressEvent = (HubEvent<Float>) hubEvent;
+                Float progress = progressEvent.getData();
+                if (progress != null && progress > 0) {
+                    op.pause();
+                }
+            }
+        });
+        subscriptions.add(pauseToken);
+
+        // Listen to Hub events to resume when operation has been paused
+        SubscriptionToken resumeToken = Amplify.Hub.subscribe(HubChannel.STORAGE, hubEvent -> {
+            if ("uploadState".equals(hubEvent.getName())) {
+                HubEvent<String> stateEvent = (HubEvent<String>) hubEvent;
+                TransferState state = TransferState.getState(stateEvent.getData());
+                if (TransferState.PAUSED.equals(state)) {
+                    Amplify.Hub.unsubscribe(pauseToken); // So it doesn't pause on each progress report
+                    Sleep.milliseconds(SLEEP_DURATION_IN_MILLISECONDS); // TODO: This is kind of gross
```

Comment:
Is this so that you can be sure and get a progress event with a time in the future? I agree its kind of gross but it'll be helpful for us to remember why it's here too

Reply:
I described it at the top when declaring the constant variable! maybe i should copy and paste the comment here too

Reply:
Ah okay, I just missed it. Once is good enough, I just need to pay more attention. :-D